File: backend/app/services/billing/plans.py
from __future__ import annotations
from dataclasses import dataclass
from typing import Optional
import os
from sqlmodel import Session
from app.db.models import GlobalBillingSettings

SUPPORTED_LANGS = ["pl", "en-us", "en-uk", "fr", "de", "it", "es"]

@dataclass(frozen=True)
class Plan:
    code: str
    tier: str
    interval: str  # month/year
    price_amount: int
    currency: str
    monthly_quote_limit: Optional[int]
    language_mode: str  # single/all
    logo_allowed: bool
    show_branding: bool
    popular: bool = False

PLANS: dict[str, Plan] = {
    "DEMO": Plan(code="DEMO", tier="demo", interval="month", price_amount=0, currency="EUR", monthly_quote_limit=None, language_mode="all", logo_allowed=False, show_branding=True),
    "STARTER_M": Plan(code="STARTER_M", tier="starter", interval="month", price_amount=899, currency="EUR", monthly_quote_limit=35, language_mode="single", logo_allowed=False, show_branding=True),
    "STARTER_Y": Plan(code="STARTER_Y", tier="starter", interval="year", price_amount=8999, currency="EUR", monthly_quote_limit=35, language_mode="single", logo_allowed=False, show_branding=True),
    "PRO_M": Plan(code="PRO_M", tier="pro", interval="month", price_amount=2199, currency="EUR", monthly_quote_limit=150, language_mode="single", logo_allowed=True, show_branding=True, popular=True),
    "PRO_Y": Plan(code="PRO_Y", tier="pro", interval="year", price_amount=21999, currency="EUR", monthly_quote_limit=150, language_mode="single", logo_allowed=True, show_branding=True, popular=True),
    "PRO_M_TRIAL": Plan(code="PRO_M_TRIAL", tier="pro", interval="month", price_amount=1499, currency="EUR", monthly_quote_limit=150, language_mode="single", logo_allowed=True, show_branding=True, popular=True),
    "PRO_Y_PROMO": Plan(code="PRO_Y_PROMO", tier="pro", interval="year", price_amount=21999, currency="EUR", monthly_quote_limit=150, language_mode="single", logo_allowed=True, show_branding=True, popular=True),
    "MAX_M": Plan(code="MAX_M", tier="max", interval="month", price_amount=3999, currency="EUR", monthly_quote_limit=None, language_mode="all", logo_allowed=True, show_branding=False),
    "MAX_Y": Plan(code="MAX_Y", tier="max", interval="year", price_amount=39999, currency="EUR", monthly_quote_limit=None, language_mode="all", logo_allowed=True, show_branding=False),
}

DEFAULT_PLAN_CODE = "PRO_M"

FX_RATES = {
    "EUR": 1.0,
    "PLN": float(os.getenv("FX_RATE_PLN", "4.3")),
    "USD": float(os.getenv("FX_RATE_USD", "1.1")),
    "GBP": float(os.getenv("FX_RATE_GBP", "0.86")),
}

CURRENCY_BY_LANG = {
    "pl": "PLN",
    "en-us": "USD",
    "en-uk": "GBP",
}


def normalize_lang(lang: str | None) -> str:
    value = (lang or "").strip().lower()
    if value == "en":
        return "en-us"
    return value


def get_plan(code: str | None) -> Plan:
    if not code:
        return PLANS[DEFAULT_PLAN_CODE]
    return PLANS.get(code, PLANS[DEFAULT_PLAN_CODE])

def has_plan(code: str | None) -> bool:
    return bool(code) and code in PLANS

def currency_for_lang(lang: str | None) -> str:
    normalized = normalize_lang(lang)
    return CURRENCY_BY_LANG.get(normalized, "EUR")

def get_fx_rates(session: Session | None = None) -> dict[str, float]:
    rates = dict(FX_RATES)
    if not session:
        return rates
    settings_row = session.get(GlobalBillingSettings, 1)
    if not settings_row or not settings_row.fx_rates:
        return rates
    for key, value in settings_row.fx_rates.items():
        try:
            rate = float(value)
        except (TypeError, ValueError):
            continue
        if rate > 0:
            rates[key.upper()] = rate
    return rates

def convert_cents(amount_cents_eur: int, currency: str, rates: dict[str, float] | None = None) -> int:
    table = rates or FX_RATES
    rate = table.get(currency, 1.0)
    return int(round(amount_cents_eur * rate))

def monthly_equivalent(plan_code: str) -> float:
    plan = get_plan(plan_code)
    if plan.interval == "year":
        return plan.price_amount / 12.0
    return float(plan.price_amount)

def allowed_languages(plan: Plan, tenant_lang: str | None) -> list[str]:
    if plan.language_mode == "all":
        return SUPPORTED_LANGS[:]
    return [normalize_lang(tenant_lang)] if tenant_lang else []

File: frontend/lib/plans.ts
import type { Lang } from "./i18n";

export type BillingInterval = "month" | "year";
export type PlanId = "starter" | "pro" | "max";

export type Plan = {
  id: PlanId;
  monthlyPriceEur: number;
  yearlyPriceEur: number;
  promoMonthlyPriceEur?: number;
  promoYearlyPriceEur?: number;
  monthlyLimit: number | null;
  languagesIncluded?: number;
  languageMode: "single" | "all";
  logoAllowed: boolean;
  showBranding: boolean;
  recommended?: boolean;
  monthlyCode: string;
  yearlyCode: string;
};

const FX_RATES = {
  EUR: 1,
  PLN: safeRate(process.env.NEXT_PUBLIC_FX_RATE_PLN, 4.3),
  USD: safeRate(process.env.NEXT_PUBLIC_FX_RATE_USD, 1.1),
  GBP: safeRate(process.env.NEXT_PUBLIC_FX_RATE_GBP, 0.86),
};

export const PLANS: Plan[] = [
  {
    id: "starter",
    monthlyPriceEur: 8.99,
    yearlyPriceEur: 89.99,
    monthlyLimit: 35,
    languageMode: "single",
    logoAllowed: false,
    showBranding: true,
    monthlyCode: "STARTER_M",
    yearlyCode: "STARTER_Y",
  },
  {
    id: "pro",
    monthlyPriceEur: 21.99,
    yearlyPriceEur: 219.99,
    promoMonthlyPriceEur: 14.99,
    monthlyLimit: 150,
    languageMode: "single",
    logoAllowed: true,
    showBranding: true,
    recommended: true,
    monthlyCode: "PRO_M_TRIAL",
    yearlyCode: "PRO_Y_PROMO",
  },
  {
    id: "max",
    monthlyPriceEur: 39.99,
    yearlyPriceEur: 399.99,
    monthlyLimit: null,
    languagesIncluded: 6,
    languageMode: "all",
    logoAllowed: true,
    showBranding: false,
    monthlyCode: "MAX_M",
    yearlyCode: "MAX_Y",
  },
];

export function planCode(planId: PlanId, interval: BillingInterval): string {
  const plan = PLANS.find((p) => p.id === planId);
  if (!plan) return "PRO_M_TRIAL";
  return interval === "year" ? plan.yearlyCode : plan.monthlyCode;
}

export function planCodeStandard(planId: PlanId, interval: BillingInterval): string {
  if (planId === "starter") return interval === "year" ? "STARTER_Y" : "STARTER_M";
  if (planId === "pro") return interval === "year" ? "PRO_Y" : "PRO_M";
  return interval === "year" ? "MAX_Y" : "MAX_M";
}

export function currencyForLang(lang: Lang): "EUR" | "PLN" | "USD" | "GBP" {
  if (lang === "pl") return "PLN";
  if (lang === "en-uk") return "GBP";
  if (lang === "en-us" || lang === "en") return "USD";
  return "EUR";
}

export function formatPrice(lang: Lang, amount: number, currency?: "EUR" | "PLN" | "USD" | "GBP"): string {
  const localeMap: Record<string, string> = {
    pl: "pl-PL",
    "en-us": "en-US",
    "en-uk": "en-GB",
    en: "en-US",
    fr: "fr-FR",
    de: "de-DE",
    it: "it-IT",
    es: "es-ES",
  };
  const locale = localeMap[lang] || "en-US";
  const curr = currency || currencyForLang(lang);
  return new Intl.NumberFormat(locale, { style: "currency", currency: curr, minimumFractionDigits: 2 }).format(amount);
}

export function displayPrice(plan: Plan, interval: BillingInterval, lang: Lang): number {
  const amountEur = interval === "year"
    ? plan.promoYearlyPriceEur ?? plan.yearlyPriceEur
    : plan.promoMonthlyPriceEur ?? plan.monthlyPriceEur;
  return convertFromEur(amountEur, lang);
}

export function regularPrice(plan: Plan, interval: BillingInterval, lang: Lang): number {
  const amountEur = interval === "year" ? plan.yearlyPriceEur : plan.monthlyPriceEur;
  return convertFromEur(amountEur, lang);
}

export function yearlySavings(plan: Plan, lang: Lang): number {
  const yearlyDisplayEur = plan.promoYearlyPriceEur ?? plan.yearlyPriceEur;
  const savingsEur = plan.monthlyPriceEur * 12 - yearlyDisplayEur;
  return convertFromEur(savingsEur, lang);
}

export function perQuotePrice(plan: Plan, interval: BillingInterval, lang: Lang): number | null {
  if (!plan.monthlyLimit) return null;
  const amountEur = interval === "year"
    ? (plan.promoYearlyPriceEur ?? plan.yearlyPriceEur) / 12
    : plan.promoMonthlyPriceEur ?? plan.monthlyPriceEur;
  return convertFromEur(amountEur / plan.monthlyLimit, lang);
}

export const LANGUAGE_OPTIONS: { value: Lang; label: string }[] = [
  { value: "pl", label: "PL" },
  { value: "en-us", label: "EN (US)" },
  { value: "en-uk", label: "EN (UK)" },
  { value: "fr", label: "FR" },
  { value: "de", label: "DE" },
  { value: "it", label: "IT" },
  { value: "es", label: "ES" },
];

function convertFromEur(amountEur: number, lang: Lang): number {
  const currency = currencyForLang(lang);
  const rate = FX_RATES[currency] || 1;
  return amountEur * rate;
}

function safeRate(raw: string | undefined, fallback: number): number {
  const parsed = Number(raw);
  if (!Number.isFinite(parsed) || parsed <= 0) return fallback;
  return parsed;
}
